external-crates/move/crates/move-analyzer/trace-adapter/src/runtime.ts

```diff
@@ -463,250 +567,329 @@ export class Runtime extends EventEmitter {
+                const stackHeight = moveCallStack.frames.length;
+                if (stackHeight <= 0) {
+                    // this should never happen
+                    throw new Error('No frame on the stack when processing Instruction event when stepping');
                 }
-            }
-            this.sendEvent(RuntimeEvents.stopOnStep);
-            return ExecutionResult.Ok;
-        } else if (currentEvent.type === TraceEventKind.ReplaceInlinedFrame) {
-            let currentFrame = this.frameStack.frames.pop();
-            if (!currentFrame) {
-                throw new Error('No frame to pop when processing `ReplaceInlinedFrame` event');
-            }
-            currentFrame.srcFileHash = currentEvent.fileHash;
-            currentFrame.optimizedSrcLines = currentEvent.optimizedLines;
-            const currentFile = this.filesMap.get(currentFrame.srcFileHash);
-            if (!currentFile) {
-                throw new Error('Cannot find file with hash '
-                    + currentFrame.srcFileHash
-                    + ' when processing `ReplaceInlinedFrame` event');
-            }
-            currentFrame.srcFilePath = currentFile.path;
-            this.frameStack.frames.push(currentFrame);
-            return this.step(next, stopAtCloseFrame);
-        } else if (currentEvent.type === TraceEventKind.OpenFrame) {
-            // if function is native then the next event will be CloseFrame
-            if (currentEvent.isNative) {
-                // see if native function aborted
-                if (this.trace.events.length > this.eventIndex + 1) {
-                    const nextEvent = this.trace.events[this.eventIndex + 1];
-                    if (nextEvent.type === TraceEventKind.Effect &&
-                        nextEvent.effect.type === TraceEffectKind.ExecutionError) {
-                        this.sendEvent(RuntimeEvents.stopOnException, nextEvent.effect.msg);
-                        return ExecutionResult.Exception;
+                const currentFrame = moveCallStack.frames[stackHeight - 1];
+                // remember last call instruction line before it (potentially) changes
+                // in the `instruction` call below
+                const lastCallInstructionLine = currentFrame.disassemblyModeTriggered
+                    ? currentFrame.lastCallInstructionBcodeLine
+                    : currentFrame.lastCallInstructionSrcLine;
+                let [sameLine, currentLine] = this.instruction(currentFrame, currentEvent);
+                // do not attempt to skip events on the same line if the previous event
+                // was a switch to/from an inlined frame - we want execution to stop before
+                // the first instruction of the inlined frame is processed
+                const prevEvent = this.trace.events[this.eventIndex - 1];
+                sameLine = sameLine &&
+                    !(prevEvent.type === TraceEventKind.ReplaceInlinedFrame
```

Comment:
This is an actual change with respect to previous implementation. As it appears frame IDs cannot be negative so previous check wouldn't work (and was incorrect anyway after adding more "artificial" frame ID values)